pkg/auth/policy.go

```diff
@@ -0,0 +1,100 @@
+package auth
+
+import (
+	"context"
+	"embed"
+	"errors"
+	"io"
+	"io/fs"
+	"net/http"
+	"strings"
+
+	"github.com/open-policy-agent/opa/loader"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/open-policy-agent/opa/topdown"
+	"github.com/rs/zerolog"
+	"github.com/rs/zerolog/log"
+	"github.com/samber/lo"
+	"golang.org/x/exp/maps"
+)
+
+// The name of the rule that must be `true` for the authorization provider to
+// permit access. This is typically provided by a policy with package name
+// `bacalhau.authz` and then by defining a rule `allow`. See
+// `policy_test_allow.rego` for a minimal example.
+const AuthzAllowRule = "bacalhau.authz.allow"
+
+type policyAuthorizer struct {
+	allowQuery rego.PreparedEvalQuery
+}
+
+type regoOpt = func(*rego.Rego)
+
+//go:embed *.rego
+var policies embed.FS
+
+// PolicyAuthorizer can authorize users by calling out to an external Rego
+// policy containing logic to make decisions about who should be authorized.
+func NewPolicyAuthorizer(policySource fs.FS, policyPath string) (Authorizer, error) {
+	results, err := loader.NewFileLoader().WithFS(policySource).All([]string{policyPath})
+	if err != nil {
+		return nil, err
+	}
+
+	opts := []regoOpt{
+		rego.Query("data." + AuthzAllowRule),
+	}
+	modules := lo.Map(maps.Values(results.Modules), func(m *loader.RegoFile, _ int) regoOpt { return rego.ParsedModule(m.Parsed) })
+	query := rego.New(append(opts, modules...)...)
+
+	allowQuery, err := query.PrepareForEval(context.TODO())
+	return &policyAuthorizer{allowQuery: allowQuery}, err
+}
+
+// ShouldAllow runs the loaded policy and provides a structure representing the
+// inbound HTTP request as input.
+func (authorizer *policyAuthorizer) ShouldAllow(req *http.Request) error {
+	if req.URL == nil {
+		return errors.New("bad HTTP request: missing URL")
+	}
+
+	body, err := io.ReadAll(req.Body)
+	if err != nil {
+		return err
+	}
+
+	in := map[string]interface{}{
+		"http": map[string]interface{}{
+			"host":    req.Host,
+			"method":  req.Method,
+			"path":    strings.Split(strings.TrimLeft(req.URL.Path, "/"), "/"),
+			"query":   req.URL.Query(),
+			"headers": req.Header,
+			"body":    string(body),
```

Comment:
The test failures seem to be happening because there is no body in the request when it gets to the endpoint (e.g. it is empty).  If I remove this read of body then it is there at the endpoint, which suggests it is being totally consumed here before it gets to the endpoint.